keras_nlp/models/gpt_neo_x/gpt_neo_x_decoder.py

```diff
@@ -0,0 +1,219 @@
+
+    def _build(self, input_shape):
+        # Create layers based on input shape.
+        self._built = True
+        self._input_shape = input_shape
+        # Infer the dimension of our hidden feature size from the build shape.
+        hidden_dim = input_shape[-1]
+
+        # Self attention layers.
+        self._self_attention_layer = GPTNeoXAttention(
+            num_heads=self.num_heads,
+            hidden_dim=hidden_dim,
+            dropout=self.dropout,
+            rotary_percentage=self.rotary_percentage,
+            rotary_max_wavelength=self.rotary_max_wavelength,
+            max_sequence_length=self.max_sequence_length,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+
+        self._self_attention_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon,
+        )
+
+        self._self_attention_dropout = keras.layers.Dropout(
+            rate=self.dropout,
+        )
+
+        # Feedforward layers.
+        self._feedforward_intermediate_dense = keras.layers.Dense(
+            self.intermediate_dim,
+            activation=self.activation,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+        self._feedforward_output_dense = keras.layers.Dense(
+            hidden_dim,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+
+        self._feedforward_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon,
+        )
+
+        self._feedforward_dropout = keras.layers.Dropout(
+            rate=self.dropout,
+        )
+
+    def call(
+        self,
+        decoder_sequence,
+        decoder_padding_mask=None,
+        decoder_attention_mask=None,
+    ):
+        if not self._built:
+            self._build(decoder_sequence.shape)
+
+        x = decoder_sequence  # Intermediate result.
+
+        # Compute self attention mask.
+        batch_size = tf.shape(decoder_sequence)[0]
+        input_length = output_length = tf.shape(decoder_sequence)[1]
+
+        self_attention_mask = compute_causal_mask(
+            batch_size, input_length, output_length, 0
+        )
+        decoder_mask = merge_padding_and_attention_mask(
+            decoder_sequence, decoder_padding_mask, decoder_attention_mask
+        )
+        if decoder_mask is not None:
+            self_attention_mask = tf.minimum(decoder_mask, self_attention_mask)
+
+        x = self._self_attention_layernorm(x)
+
+        # Self attention block.
+        x = self._self_attention_layer(
+            hidden_states=x,
+            attention_mask=self_attention_mask,
+        )
+        x = self._self_attention_dropout(x)
+
+        residual = x
```

Comment:
I wouldn't call this a residual because the decoder_sequence is really the residual in this case, while your feedforward and attention are computed in parallel.

```python
residual = decoder_sequence

# Compute attention and feedforward in parallel (parallel attention).
x = self._self_attention_layernorm(decoder_sequence)
x = self._self_attention_layer(
    hidden_states=x,
    attention_mask=self_attention_mask,
)
x = self._self_attention_dropout(x)
attention_output = x

x = self._feedforward_layernorm(decoder_sequence)
x = self._feedforward_intermediate_dense(x)
x = self._feedforward_output_dense(x)
feedforward_output = x
return feedforward_output + attention_output + residual
```